Source/Core/Core/IOS/USB/Emulated/Infinity.cpp

```diff
@@ -0,0 +1,907 @@
+      m_response_list.pop();
+    }
+  }
+
+  return 0;
+}
+int InfinityUSB::SubmitTransfer(std::unique_ptr<IsoMessage> cmd)
+{
+  DEBUG_LOG_FMT(IOS_USB,
+                "[{:04x}:{:04x} {}] Isochronous: length={:04x} endpoint={:02x} num_packets={:02x}",
+                m_vid, m_pid, m_active_interface, cmd->length, cmd->endpoint, cmd->num_packets);
+  return 0;
+}
+void InfinityUSB::ScheduleTransfer(std::unique_ptr<TransferCommand> command,
+                                   const std::array<u8, 32>& data, s32 expected_count,
+                                   u64 expected_time_us)
+{
+  command->FillBuffer(data.data(), expected_count);
+  command->ScheduleTransferCompletion(expected_count, expected_time_us);
+}
+
+bool InfinityBase::HasFigureBeenAddedRemoved() const
+{
+  return !m_figure_added_removed_response.empty();
+}
+
+std::array<u8, 32> InfinityBase::PopAddedRemovedResponse()
+{
+  std::array<u8, 32> response = m_figure_added_removed_response.front();
+  m_figure_added_removed_response.pop();
+  return response;
+}
+
+u8 InfinityBase::GenerateChecksum(const std::array<u8, 32>& data, int num_of_bytes) const
+{
+  int checksum = 0;
+  for (int i = 0; i < num_of_bytes; i++)
+  {
+    checksum += data[i];
+  }
+  return (checksum & 0xFF);
+}
+
+void InfinityBase::GetBlankResponse(u8 sequence, std::array<u8, 32>& reply_buf)
+{
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x01;
+  reply_buf[2] = sequence;
+  reply_buf[3] = GenerateChecksum(reply_buf, 3);
+}
+
+void InfinityBase::GetPresentFigures(u8 sequence, std::array<u8, 32>& reply_buf)
+{
+  int x = 3;
+  for (u8 i = 0; i < m_figures.size(); i++)
+  {
+    u8 slot = i == 0 ? 0x10 : (i > 0 && i < 4) ? 0x20 : 0x30;
+    if (m_figures[i].present)
+    {
+      reply_buf[x] = slot + m_figures[i].order_added;
+      reply_buf[x + 1] = 0x09;
+      x = x + 2;
+    }
+  }
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = x - 2;
+  reply_buf[2] = sequence;
+  reply_buf[x] = GenerateChecksum(reply_buf, x);
+}
+
+void InfinityBase::GetFigureIdentifier(u8 fig_num, u8 sequence, std::array<u8, 32>& reply_buf)
+{
+  std::lock_guard lock(m_infinity_mutex);
+
+  InfinityFigure& figure = GetFigureByOrder(fig_num);
+
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x09;
+  reply_buf[2] = sequence;
+  reply_buf[3] = 0x00;
+
+  if (figure.present)
+  {
+    memcpy(&reply_buf[4], figure.data.data(), 7);
+  }
+  reply_buf[11] = GenerateChecksum(reply_buf, 11);
+}
+
+void InfinityBase::QueryBlock(u8 fig_num, u8 block, std::array<u8, 32>& reply_buf, u8 sequence)
+{
+  std::lock_guard lock(m_infinity_mutex);
+
+  InfinityFigure& figure = GetFigureByOrder(fig_num);
+
+  reply_buf[0] = 0xaa;
+  reply_buf[1] = 0x12;
+  reply_buf[2] = sequence;
+  reply_buf[3] = 0x00;
+  u8 file_block = 0;
+  if (block == 0)
+  {
+    file_block = 1;
+  }
+  else
+  {
+    file_block = block * 4;
+  }
+  if (figure.present)
+  {
+    memcpy(&reply_buf[4], figure.data.data() + (16 * (file_block)), 16);
```

Comment:
Please add a bounds check for `file_block`. Also, the parens around file_block are unnecessary.